test/performance/core/simd/view_to_simd_chunk_benchmark.cpp

```diff
@@ -0,0 +1,120 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <deque>
+#include <iterator>
+#include <list>
+#include <vector>
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+#include <seqan3/core/simd/concept.hpp>
+#include <seqan3/core/simd/simd_traits.hpp>
+#include <seqan3/core/simd/simd.hpp>
+#include <seqan3/core/simd/view_to_simd.hpp>
+#include <seqan3/range/container/aligned_allocator.hpp>
+#include <seqan3/std/ranges>
+#include <seqan3/test/performance/sequence_generator.hpp>
+
+using namespace seqan3;
+
+// ============================================================================
+//  naive implementation
+// ============================================================================
+
+template <typename container_t, typename simd_t>
+void to_simd_naive(benchmark::State& state)
+{
+    // Preparing the sequences
+    std::vector<container_t> data;
+    data.resize(simd_traits<simd_t>::length);
+
+    for (size_t i = 0; i < simd_traits<simd_t>::length; ++i)
+        std::ranges::copy(test::generate_sequence<dna4>(500, 10), std::back_inserter(data[i]));
+
+    std::vector<simd_t, aligned_allocator<simd_t, sizeof(simd_t)>> v;
+
+    size_t max_size = 0;
+    for (auto & seq : data)
+        max_size = std::max(max_size, seq.size());
+
+    v.resize(max_size);
+
+    size_t value = 0;
+    for (auto _ : state)
+    {
+        for (size_t i = 0; i < max_size; ++i)
+            for (size_t j = 0; j < simd_traits<simd_t>::length; ++j)
+                v[i][j] = (i < data[j].size()) ? seqan3::to_rank(data[j][i]) : 0;
```

Comment:
I could resize the sequences before to the max_size and fill them with the padding value and transform them, which is not comparable to the view implementation since it should not modify the input sequences. I could also split the loop into the min size of the sequences such that within in the inner loop all access is guaranteed to work. Might be worth it.

Reply:
```
------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                    Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------------------------
// new naive code without condition in hot-loop
to_simd_naive_wo_condition<std::vector<dna4>, simd_type_t<int8_t>>        9328 ns         9292 ns        74052 value=55.6871M
to_simd_naive_wo_condition<std::vector<dna4>, simd_type_t<int16_t>>       3797 ns         3753 ns       195680 value=147.151M
to_simd_naive_wo_condition<std::vector<dna4>, simd_type_t<int32_t>>       1846 ns         1840 ns       373842 value=281.129M
to_simd_naive_wo_condition<std::vector<dna4>, simd_type_t<int64_t>>       1256 ns         1251 ns       554957 value=417.328M
to_simd_naive_wo_condition<std::deque<dna4>, simd_type_t<int8_t>>        15799 ns        15768 ns        44143 value=33.1955M
to_simd_naive_wo_condition<std::deque<dna4>, simd_type_t<int16_t>>        5952 ns         5927 ns       114995 value=86.4762M
to_simd_naive_wo_condition<std::deque<dna4>, simd_type_t<int32_t>>        3193 ns         3187 ns       223111 value=167.779M
to_simd_naive_wo_condition<std::deque<dna4>, simd_type_t<int64_t>>        2031 ns         2021 ns       341016 value=256.444M

// Old naive code
to_simd_naive_w_condition<std::vector<dna4>, simd_type_t<int8_t>>         6931 ns         6914 ns        99541 value=74.8548M
to_simd_naive_w_condition<std::vector<dna4>, simd_type_t<int16_t>>        2315 ns         2307 ns       290175 value=218.212M
to_simd_naive_w_condition<std::vector<dna4>, simd_type_t<int32_t>>        1334 ns         1327 ns       536300 value=403.298M
to_simd_naive_w_condition<std::vector<dna4>, simd_type_t<int64_t>>        1065 ns         1062 ns       665602 value=500.533M
to_simd_naive_w_condition<std::deque<dna4>, simd_type_t<int8_t>>         26413 ns        26271 ns        26946 value=20.2634M
to_simd_naive_w_condition<std::deque<dna4>, simd_type_t<int16_t>>         7931 ns         7910 ns        87900 value=66.1008M
to_simd_naive_w_condition<std::deque<dna4>, simd_type_t<int32_t>>         2739 ns         2728 ns       254276 value=191.216M
to_simd_naive_w_condition<std::deque<dna4>, simd_type_t<int64_t>>         1560 ns         1557 ns       449710 value=338.182M

// to simd view
to_simd<std::vector<dna4>, simd_type_t<int8_t>>                           2320 ns         2314 ns       302766 value=241.002M
to_simd<std::vector<dna4>, simd_type_t<int16_t>>                          1503 ns         1499 ns       473066 value=361.422M
to_simd<std::vector<dna4>, simd_type_t<int32_t>>                          1309 ns         1305 ns       534359 value=408.25M
to_simd<std::vector<dna4>, simd_type_t<int64_t>>                           872 ns          864 ns       849216 value=642.007M
to_simd<std::deque<dna4>, simd_type_t<int8_t>>                            9016 ns         8988 ns        77299 value=61.53M
to_simd<std::deque<dna4>, simd_type_t<int16_t>>                           6047 ns         6035 ns       114435 value=87.4283M
to_simd<std::deque<dna4>, simd_type_t<int32_t>>                           2266 ns         2263 ns       308044 value=235.346M
to_simd<std::deque<dna4>, simd_type_t<int64_t>>                           1238 ns         1232 ns       577930 value=436.915M
to_simd<std::list<dna4>, simd_type_t<int8_t>>                            13909 ns        13829 ns        50849 value=40.4758M
to_simd<std::list<dna4>, simd_type_t<int16_t>>                            4384 ns         4371 ns       135533 value=103.547M
to_simd<std::list<dna4>, simd_type_t<int32_t>>                            1521 ns         1512 ns       434518 value=331.972M
to_simd<std::list<dna4>, simd_type_t<int64_t>>                             922 ns          919 ns       784094 value=592.775M
```

Reply:
So here are new benchmarks after I added the naive implementation without a condition inside of the hot loop. That was the least complex implementation I could think of. While this implementation seems better for non contiguous ranges it is always worst than the to_simd view.

Reply:
Maybe you have a better implementation for it? We can then try to adapt and replace it once this code is merged.

Reply:
thank you for testing